File: go/1161.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func maxLevelSum(root *TreeNode) int {
	queue := make([]*TreeNode, 0)
	queue = append(queue, root)
	ret, max := 0, -(2 << 31)

	for level := 1; len(queue) > 0; level++ {
		sum := 0
		for size := len(queue); size > 0; size-- {
			head := queue[0]
			queue = queue[1:]

			sum += head.Val

			if head.Left != nil {
				queue = append(queue, head.Left)
			}
			if head.Right != nil {
				queue = append(queue, head.Right)
			}
		}
		if sum > max {
			max, ret = sum, level
		}
	}

	return ret
}

File: cpp/1161.cpp
#include <climits>
#include <cstddef>
#include <queue>

// Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};

class Solution {
public:
  int maxLevelSum(TreeNode *root) {
    size_t ret{0}, level{1};
    int max{INT_MIN};

    std::queue<TreeNode *> q;
    q.push(root);
    while (q.size()) {
      size_t size{q.size()};
      int sum{0};
      while (size) {
        TreeNode *cur{q.front()};
        q.pop();
        sum += cur->val;
        --size;
        if (cur->left)
          q.push(cur->left);
        if (cur->right)
          q.push(cur->right);
      }
      if (sum > max)
        max = sum, ret = level;
      ++level;
    }

    return ret;
  }
};
